### PR TITLE
WIP: upgrade faust from qt4 to qt5

### PR DESCRIPTION
**Pull request by magnetophon**
###### Description of changes
This fixes https://github.com/NixOS/nixpkgs/pull/174634 for faust2alqt and faust2jaqt.
These are compilation scripts, that spit out a program that uses qt.
I changed the script so it wraps the output program so it knows QT_PLUGIN_PATH.
This solution works, but is kinda clunky: you end up with two files, one pointing at the other, which is hidden.
Thanks to @tpwrules who pointed out this solution on IRC 

faust2lv2 on the other hand, spits out an lv2 plugin, to be used inside a DAW, like ardour.
I do not see how I could make the DAW find QT_PLUGIN_PATH.
Any ideas?

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/faust/faust2alqt.nix [hunk 1]
```diff
@@ -1,15 +1,39 @@
...
   ];
 
+  dontWrapQtApps = true;
+
+  # Wrap the binary coming out of the the compilation script, so it knows QT_PLUGIN_PATH
+  wrapBinary = writeText "wrapBinary" ''
+    for p in $FILES; do
+      binary=$(basename --suffix=.dsp "$p")
+      touch "$binary"-temp-wrap.sh
+      echo '#!/usr/bin/env bash' >> "$binary"-temp-wrap.sh
```
Comment (ajs124) [inline pkgs/applications/audio/faust/faust2alqt.nix:23]: could use `pkgs.runtimeShell`

**Review by ajs124 — COMMENTED** [review]
maybe you can somehow incorporate `makeBinaryWrapper` or `makeWrapper` into this, instead of rolling your own? not sure if there's a good way to do that :thinking:

**Comment by magnetophon** [comment]
> maybe you can somehow incorporate `makeBinaryWrapper` or `makeWrapper` into this, instead of rolling your own? not sure if there's a good way to do that thinking

That is what I tried first, but I have not found a way to call these at runtime.

**Comment by ajs124** [comment]
> > maybe you can somehow incorporate `makeBinaryWrapper` or `makeWrapper` into this, instead of rolling your own? not sure if there's a good way to do that thinking
> 
> That is what I tried first, but I have not found a way to call these at runtime.

Would something like this work?
```diff
--- a/pkgs/applications/audio/faust/faust2alqt.nix
+++ b/pkgs/applications/audio/faust/faust2alqt.nix
@@ -2,7 +2,7 @@
 , alsa-lib
 , qtbase
 , writeText
-, runtimeShell
+, makeBinaryWrapper
 }:
 
 faust.wrapWithBuildEnv rec {
@@ -18,15 +18,10 @@ faust.wrapWithBuildEnv rec {
 
   # Wrap the binary coming out of the the compilation script, so it knows QT_PLUGIN_PATH
   wrapBinary = writeText "wrapBinary" ''
+    source ${makeBinaryWrapper}/nix-support/hook
+
     for p in $FILES; do
-      binary=$(basename --suffix=.dsp "$p")
-      touch "$binary"-temp-wrap.sh
-      echo '#!${runtimeShell}' >> "$binary"-temp-wrap.sh
-      echo 'export QT_PLUGIN_PATH=${qtbase}/${qtbase.qtPluginPrefix}' >> "$binary"-temp-wrap.sh
-      echo "./."$binary"-wrapped" >> "$binary"-temp-wrap.sh
-      mv "$binary" ."$binary"-wrapped
-      mv "$binary"-temp-wrap.sh "$binary"
-      chmod +x "$binary"
+      makeWrapper --set QT_PLUGIN_PATH "${qtbase}/${qtbase.qtPluginPrefix}"
     done
   '';
 
```

> Do you have any idea how I could get `faust2lv2` working with `qt5`?

No idea. I don't really know how to test any of these scripts, tbh.

**Comment by magnetophon** [comment]
Thanks for the idea, but I couldn't get it working.
After I a few iterations I ended up with:
```nix
  wrapBinary = writeText "wrapBinary" ''
    source ${dieHook}/nix-support/setup-hook
    source ${makeBinaryWrapper}/nix-support/setup-hook
    for p in $FILES; do
      binary=$(basename --suffix=.dsp "$p")
      makeWrapper $binary $binary --set QT_PLUGIN_PATH "${qtbase}/${qtbase.qtPluginPrefix}"
    done
  '';
```
But that errors out on:
```
/nix/store/4aqk6dgyqr0h55s237yl0nvyyqf4awn6-hook/nix-support/setup-hook: line 98: main: unbound variable
<stdin>:2: error: ISO C forbids an empty translation unit [-Werror=pedantic]
cc1: all warnings being treated as errors
```
Any ideas?

Without ``source ${dieHook}/nix-support/setup-hook`` it complains about missing the function ``die``.

**Comment by magnetophon** [comment]
@ajs124 OK, done.
When you think this PR is ready, I'll squash the commits.

File: pkgs/applications/audio/faust/faust2alqt.nix [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
   ];
 
+  dontWrapQtApps = true;
+
+  # Wrap the binary coming out of the the compilation script, so it knows QT_PLUGIN_PATH
+  wrapBinary = writeText "wrapBinary" ''
+    source ${makeWrapper}/nix-support/setup-hook
+    for p in $FILES; do
+      binary=$(basename --suffix=.dsp "$p")
+      mv $binary ."$binary"-wrapped
```
Comment (ajs124) [inline pkgs/applications/audio/faust/faust2alqt.nix:24]: doesn't makeWrapper do all of this moving around on its own?

File: pkgs/applications/audio/faust/faust2alqt.nix [hunk 3]
```diff
@@ -1,15 +1,37 @@
...
-    qt4
+    qtbase
   ];
 
+  dontWrapQtApps = true;
+
+  # Wrap the binary coming out of the the compilation script, so it knows QT_PLUGIN_PATH
+  wrapBinary = writeText "wrapBinary" ''
+    source ${makeWrapper}/nix-support/setup-hook
+    for p in $FILES; do
```
Comment (ajs124) [inline pkgs/applications/audio/faust/faust2alqt.nix:22]: skimming the script this gets appended to, I saw something like `$BINARIES`. why are we using FILES and not that?

**Review by ajs124 — COMMENTED** [review]
How do I test this? I don't really understand this software.

File: pkgs/applications/audio/faust/faust2alqt.nix (hunk 3 above)
Comment (magnetophon) [reply pkgs/applications/audio/faust/faust2alqt.nix:22]: ``$BINARIES`` has extra characters in it that would have needed to be stripped out.
This was just simpler.

**Pull request closed without merging**